packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart

```diff
@@ -0,0 +1,290 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+library amplify_push_notifications;
+
+import 'dart:async';
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:amplify_push_notifications/src/native_push_notifications_plugin.g.dart';
+import 'package:flutter/services.dart';
+
+const _tokenReceivedEventChannel = EventChannel(
+  'com.amazonaws.amplify/push_notification/event/TOKEN_RECEIVED',
+);
+const _notificationOpenedEventChannel = EventChannel(
+  'com.amazonaws.amplify/push_notification/event/NOTIFICATION_OPENED',
+);
+const _foregroundNotificationEventChannel = EventChannel(
+  'com.amazonaws.amplify/push_notification/event/FOREGROUND_MESSAGE_RECEIVED',
+);
+
+/// {@template amplify_push_notifications.amplify_push_notifications}
+/// Implementation of the Amplify Push Notifications category.
+///
+/// - Concrete implementation of the Plugin Interface
+/// - Communicates with native module via [PushNotificationsHostApi]
+/// {@endtemplate}
+class AmplifyPushNotifications extends PushNotificationsPluginInterface {
+  /// {@macro amplify_push_notifications.amplify_push_notifications}
+  AmplifyPushNotifications({
+    required ServiceProviderClient serviceProviderClient,
+  })  : _serviceProviderClient = serviceProviderClient,
+        _hostApi = PushNotificationsHostApi(),
+        _flutterApi = _PushNotificationsFlutterApi() {
+    _flutterApi.registerOnLaunchNotificationOpenedCallback((remotePushMessage) {
+      _launchNotification = remotePushMessage;
+      _launchNotificationForAnalytics = remotePushMessage;
+    });
+
+    _onTokenReceived = _tokenReceivedEventChannel
+        .receiveBroadcastStream()
+        .cast<Map<Object?, Object?>>()
+        .map((payload) {
+      return payload['token'] as String;
+    }).distinct();
+
+    _onForegroundNotificationReceived = _foregroundNotificationEventChannel
+        .receiveBroadcastStream()
+        .cast<Map<Object?, Object?>>()
+        .map(PushNotificationMessage.fromJson);
+
+    _onNotificationOpened = _notificationOpenedEventChannel
+        .receiveBroadcastStream()
+        .cast<Map<Object?, Object?>>()
+        .map(PushNotificationMessage.fromJson);
+  }
+
+  final AmplifyLogger _logger =
+      AmplifyLogger.category(Category.pushNotifications)
+          .createChild('AmplifyPushNotification');
+  final ServiceProviderClient _serviceProviderClient;
+  final PushNotificationsHostApi _hostApi;
+  final _PushNotificationsFlutterApi _flutterApi;
+
+  late final Stream<String> _onTokenReceived;
+  late final Stream<PushNotificationMessage> _onForegroundNotificationReceived;
+  late final Stream<PushNotificationMessage> _onNotificationOpened;
+
+  var _isConfigured = false;
+  PushNotificationMessage? _launchNotification;
+  PushNotificationMessage? _launchNotificationForAnalytics;
```

Comment:
This copy will be used to send an analytic event record after `Amplify.configure` is completed, used internally.

Reply:
This should be retrieved once from the platform - if we need to use it for analytics purposes it should be used then and there. It does not need to be stored on the class.